### PR TITLE
[generate] `PromptLookupCandidateGenerator` won't generate forbidden tokens

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

See title.

In the process, deflakes a lot of tests :) (It was generating forbidden tokens in tests, like image tokens in VLMs -- now we can specify which sequences are forbidden)

`py.test tests/models/voxtral/test_modeling_voxtral.py -k test_prompt_lookup_decoding_matches_greedy_search --flake-finder --flake-runs 1000` now runs with without problems.

### CODE REVIEW

File: src/transformers/generation/candidate_generator.py [hunk 1]
```diff
@@ -1019,11 +1019,15 @@ def __init__(
         num_output_tokens: int = 10,
         max_matching_ngram_size: Optional[int] = None,
         max_length: int = 20,
+        bad_words_ids: Optional[list[list[int]]] = None,
```
Comment (ydshieh) [inline src/transformers/generation/candidate_generator.py]: maybe somewhere to explain this in more details.

I get the idea, but not clear about the format, like why it's a list of list of int, and below it is converted to list of tensors, and how this format is sued in the actual logic `if any(bad_word_id.to(chosen_ids.device) in chosen_ids for bad_word_id in self.bad_words_ids):`.

Just nit but would be nice to have

**Review by ydshieh — APPROVED** [review]
Super!!!!!

File: tests/generation/test_utils.py [hunk 2]
```diff
@@ -779,27 +779,6 @@ def test_prompt_lookup_decoding_matches_greedy_search(self):
...
-                    "ayavision",
-                    "janus",
-                    "gemma3",
-                    "mistral3",
-                    "chameleon",
-                    "internvl",
-                    "qwen2_5omni",  # the file is named `qwen2_5_omni`, but the model class is `Qwen2_5Omni`,
-                    # All models below: shouldn't suggest audio tokens. Can be fixed by passing `suppress_ids` to candidate generator: @joaa @raushan
-                    "voxtral",
-                    "qwen2audio",
```
Comment (zucchini-nlp) [inline tests/generation/test_utils.py:802]: yaaaaaay 💃🏻

File: src/transformers/generation/utils.py [hunk 3]
```diff
@@ -1025,6 +1025,7 @@ def _get_candidate_generator(
                 num_output_tokens=generation_config.prompt_lookup_num_tokens,
                 max_matching_ngram_size=generation_config.max_matching_ngram_size,
                 max_length=generation_config.max_length,
+                bad_words_ids=generation_config.bad_words_ids,
```
Comment (zucchini-nlp) [inline src/transformers/generation/utils.py]: not sure if we plan to keep PromptDecoding in the future. If we are, I think we should raise a warning that some generation config params will not be used, e.g. `no_repeat_ngram_size`

**Review by zucchini-nlp — COMMENTED** [review]
Great thanks! One question, should we raise a warning that prompt decoder will not filter candidates by logits so we can't support all generation config params?

File: src/transformers/generation/candidate_generator.py (hunk 1 above)
Comment (gante) [reply src/transformers/generation/candidate_generator.py]: Added in the docstring 🤗

**Comment by gante** [comment]
@zucchini-nlp great comment. We can indeed make it work with ANY logits processor that blocks tokens, not just `bad_word_ids`. 

The logic is slightly more complex, but it is added in the latest commit. In a nutshell, we simulate using the logits processor with fake input logits, using the same processors as we use for the main model, and any selected token with `-inf` logits is guaranteed to be a forbidden token. This also works with external logits processors 🤗 

Regarding long-term maintenance: if we're keeping assisted generation (we are for now), we're also keeping this one. This is the candidate-based generation strategy with the fewest requirements.

File: src/transformers/generation/candidate_generator.py [hunk 4]
```diff
@@ -1004,26 +1004,39 @@ class PromptLookupCandidateGenerator(CandidateGenerator):
     Read the following blog post for more information: https://github.com/apoorvumang/prompt-lookup-decoding
 
     Args:
-        max_matching_ngram_size (`int`):
-            The maximum ngram size to be considered for matching in the prompt
-        num_output_tokens (`int`):
+        eos_token_id (`torch.Tensor`, *optional*):
```
Comment (gante) [inline src/transformers/generation/candidate_generator.py:1007]: (docstring args were out of order, and some were missing)

File: src/transformers/generation/candidate_generator.py [hunk 5]
```diff
@@ -1061,13 +1074,41 @@ def get_candidates(self, input_ids: torch.LongTensor) -> tuple[torch.LongTensor,
...
+                    # tokens are allowed, we keep `chosen_ids` as is.
+                    # 1. create random logits.
+                    # 2. apply the logits processor to get output logits for the next token, using the arbitrary
+                    #    logits as input.
+                    # 3. compare the output logits with the next candidate token. If they are -inf, then the next
+                    #    candidate token is forbidden and we don't want to generate it.
+                    if self.logits_processor is not None:
+                        sequence_with_candidate = input_ids
+                        for candidate_idx, new_candidate_token in enumerate(chosen_ids):
+                            fake_input_logits = torch.ones((bsz, self.vocab_size), device=input_ids.device)
```
Comment (zucchini-nlp) [inline src/transformers/generation/candidate_generator.py]: prob we can create a fake logit once, since it's always same. Can be helpful with models like gemma with huuge vocab size

**Review by zucchini-nlp — APPROVED** [review]
Nice, happy that it worked well! Thanks for aligning prompt decoder with common generation API. Left one tiny question otherwise lgtm

File: src/transformers/generation/candidate_generator.py (hunk 5 above)
Comment (gante) [reply src/transformers/generation/candidate_generator.py]: I thought so at first, but then considered that in-place ops (in custom logits processors) might behave poorly. Double-checked it now, it seems resilient to in-place ops -> changed it.
